### PR TITLE
Refactor design-time tools

### PR DESCRIPTION
**Pull request by natemcmaster**
Split design-time tools into three packages so that users need only install "Microsoft.EntityFrameworkCore.Tools" and nuget will restore the tools they need for the platform they target.

The packages:

 - **.Tools**
     - A pseudo-metapackage. This contains shims to call into shared
       components.
     - **lib/net451/Microsoft.EntityFrameworkCore.Tools.dll** - Contains
       PowerShell cmdlets and APIs that bridge it with common components.
     - **lib/netcoreapp1.0/dotnet-ef.dll** - a light shim for initializing
       commands on DotNet project. This executes from "tools".
     - **lib/netcore50/Microsoft.EntityFrameworkCore.Tools.dll** - same as
       net451. For compatibility with UWP only. This is never actually
       executed.
     - **build/netcore50/*.target** - swaps netcore50 to net451 at build time
       on UWP projects
     - **tools/*** - the powershell cmdlets
 - **.Tools.Cli**
     - A program to launch commands from dotnet-cli. These execute as a
       "dependency".
     - **lib/netstandard1.5/*.dll** - Contains the entry point for executing
       commands on a DotNet project model that supports .NET Core
     - **lib/net451/*.exe** - Contains the entry point for executing commands
       on a DotNet project modle that supports desktop .NET only
 - **.Tools.Core**
     - contains shared components. This contains the implementation
       details of executing commands.

Resolves https://github.com/aspnet/EntityFramework/issues/4577
Resolves https://github.com/aspnet/EntityFramework/issues/5065
Partial fix for https://github.com/aspnet/EntityFramework/issues/3925

### Functional issues

- [x] Can't use dotnet-ef in a class library or standalone apps. This dies with "Could not resolve coreclr path". https://github.com/dotnet/cli/issues/2645
- [x] There are issues loading assemblies when startup project and current project are different. Haven't identified the source of the issue.
- [x] Cannot install into a csproj. Blocked on https://github.com/NuGet/Home/issues/2635.
- [x] The framework/configuration on the dispatcher (dotnet-ef) should flow through to the "inside-man".

### Cleanup

- [x] Put all strings into resx files
- [x] Remove workaround that runs `dotnet restore --infer-runtimes` on the tools projects. Without this, dotnet-pack fails, complaining that no runtimes were specified.
- [ ] Merge https://github.com/aspnet/Coherence/pull/79
 
cc @divega @NTaylorMullen @bricelam

### CODE REVIEW

File: src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+using Microsoft.DotNet.ProjectModel;
+using Microsoft.Extensions.PlatformAbstractions;
+using NuGet.Frameworks;
+using Microsoft.EntityFrameworkCore.Design;
+using Microsoft.EntityFrameworkCore.Internal;
+using Microsoft.EntityFrameworkCore.Migrations.Design;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore.Scaffolding.Internal;
+using Microsoft.EntityFrameworkCore.Design.Internal;
```
Comment (AndriySvyryd) [inline src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs]: Sort usings

**Comment by AndriySvyryd** [comment]
🐑 🇮🇹

File: src/Microsoft.EntityFrameworkCore.Tools.Cli/ConsoleCommandLogger.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        }
+
+        protected override void WriteDebug(string message)
+            => Reporter.Verbose.WriteLine(message.Bold().Black());
+
+        protected override void WriteError(string message)
+            => Reporter.Error.WriteLine(message.Bold().Red());
+
+        protected override void WriteInformation(string message)
+            => Reporter.Error.WriteLine(message);
```
Comment (NTaylorMullen) [inline src/Microsoft.EntityFrameworkCore.Tools.Cli/ConsoleCommandLogger.cs]: Error for writeinformation?

File: src/Microsoft.EntityFrameworkCore.Tools.Cli/ExecuteCommand.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+
+using JetBrains.Annotations;
+using Microsoft.Extensions.CommandLineUtils;
+using Microsoft.EntityFrameworkCore.Design;
+
+namespace Microsoft.EntityFrameworkCore.Tools.Cli
+{
+    public class ExecuteCommand
```
Comment (NTaylorMullen) [inline src/Microsoft.EntityFrameworkCore.Tools.Cli/ExecuteCommand.cs]: static

File: src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+            string startupBuildBaseDir = null;
+            string startupOutputDir = null;
+
+            var startupProjectContext = GetCompatibleStartupProjectContext(startupProject);
+
+            var startupFramework = startupProjectContext.TargetFramework;
+
+            if (buildStartup)
+            {
+                Reporter.Verbose.WriteLine("Build started...".Bold().Black());
```
Comment (NTaylorMullen) [inline src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs]: Note: We're probably gonna have them remove these string extensions from the cli.utils package. For now is fine, but an FYI

File: src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+                        startupBuildBaseDir,
+                        startupOutputDir)
+                    .ForwardStdOut()
+                    .ForwardStdErr()
+                    .Execute()
+                    .ExitCode;
+
+                if (buildExitCode != 0)
+                {
+                    throw new OperationException("Build failed.");
```
Comment (NTaylorMullen) [inline src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs]: If this is going to be consumed by VS/scaffolding may want to provide some extra information here so when those tools consume ef they provide verbose enough errors when things go wrong.

File: src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+    {
+        private const string DataDirEnvName = "ADONET_DATA_DIR";
+
+        private readonly LazyRef<MigrationsOperations> _migrationsOperations;
+        private readonly LazyRef<DatabaseOperations> _databaseOperations;
+        private readonly LazyRef<DbContextOperations> _contextOperations;
+
+        public OperationExecutor([CanBeNull] string startupProject, [CanBeNull] string environment)
+        {
+            var project = Path.Combine(Directory.GetCurrentDirectory(), Project.FileName);
```
Comment (NTaylorMullen) [inline src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs]: move this into the if statement below

File: src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+            var buildStartup = true;
+            if (startupProject == null)
+            {
+                startupProject = project;
+                buildStartup = false;
+            }
+            // TODO flow through from dispatch
```
Comment (NTaylorMullen) [inline src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs]: Wont this need to be done in the dispatch layer? Reason: Users project needs to be built in order to dispatch in the first place.

File: src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+            var runtimeOutputPath = startupProjectContext.GetOutputPaths(startupConfiguration)?.RuntimeOutputPath;
+            if (!string.IsNullOrEmpty(runtimeOutputPath))
+            {
+                Environment.SetEnvironmentVariable(DataDirEnvName, runtimeOutputPath);
+#if NET451
+                AppDomain.CurrentDomain.SetData("DataDirectory", runtimeOutputPath);
+#endif
+            }
+
+            var projectFile = ProjectReader.GetProject(project);
```
Comment (NTaylorMullen) [inline src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs]: Why does this and the startupProject bits potentially fork?

File: src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+            var startupAssemblyName = startupProjectContext.ProjectFile.GetCompilerOptions(startupFramework, startupConfiguration).OutputName;
+            var assemblyName = projectFile.GetCompilerOptions(projectFramework, projectConfiguration).OutputName;
+            var projectDir = projectFile.ProjectDirectory;
+            var startupProjectDir = startupProjectContext.ProjectFile.ProjectDirectory;
+            // TODO should this match assembly name?
+            var rootNamespace = projectFile.Name;
+
+            var assemblyLoader = new AssemblyLoader(Assembly.Load);
```
Comment (NTaylorMullen) [inline src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs]: Why not just utilize Assembly.Load?

File: src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+                cancellationToken);
+
+        private ProjectContext GetCompatibleStartupProjectContext(string projectPath)
+        {
+            var projectFile = ProjectReader.GetProject(projectPath);
+            var frameworks = projectFile.GetTargetFrameworks().Select(f => f.FrameworkName);
+
+#if NET451
+            var currentFramework = FrameworkConstants.CommonFrameworks.Net451;
+            var framework = frameworks.FirstOrDefault(f => f.IsDesktop())
```
Comment (NTaylorMullen) [inline src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs]: Based on [this](https://github.com/aspnet/EntityFramework/pull/5142/files#r60815303) having frameworks here may not be interesting.

File: src/Microsoft.EntityFrameworkCore.Tools.Cli/ExecuteCommand.cs [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    {
+        public static EfCommandLineApplication Configure()
```
Comment (NTaylorMullen) [inline src/Microsoft.EntityFrameworkCore.Tools.Cli/ExecuteCommand.cs]: `Create` would probably make more sense.

File: src/Microsoft.EntityFrameworkCore.Tools.Cli/Program.cs [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+            catch (Exception ex)
+            {
+                if (ex is TargetInvocationException)
+                {
+                    ex = ex.InnerException;
+                }
+
+                if (!(ex is OperationException))
+                {
+                    Reporter.Error.WriteLine(ex.ToString());
```
Comment (NTaylorMullen) [inline src/Microsoft.EntityFrameworkCore.Tools.Cli/Program.cs]: This'll result in some duplicated output with the reporter writeline below

File: src/Microsoft.EntityFrameworkCore.Tools/Program.cs [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+                {
+                    Reporter.Error.WriteLine(ex.ToString());
+                }
+
+                Reporter.Error.WriteLine(ex.Message.Bold().Red());
+                return 1;
+            }
+        }
+
+        private static void HandleVerboseOption(ref string[] args)
```
Comment (NTaylorMullen) [inline src/Microsoft.EntityFrameworkCore.Tools/Program.cs]: Why not rely on dotnets verbose options/Reporter.Verbose?

File: src/Microsoft.EntityFrameworkCore.Tools/DispatchCommand.cs [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+                        .ForwardStdOut()
+                        .Execute()
+                        .ExitCode;
+                    if (buildExitCode != 0)
+                    {
+                        throw new OperationException("Build failed.");
+                    }
+                }
+
+                var projectContext = new ProjectContextBuilder()
```
Comment (NTaylorMullen) [inline src/Microsoft.EntityFrameworkCore.Tools/DispatchCommand.cs]: Don't need a projectContext. See https://github.com/aspnet/RazorTooling/pull/65/files#diff-33c97b345caf0ab940bd32a46634dde0R73

**Comment by NTaylorMullen** [comment]
:watch:

File: src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs (hunk 7 above)
Comment (natemcmaster) [inline src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs]: Needs to be done in both places. Example:
```
AspNetWebApp/
    Startup.cs
Library/
    TestContext.cs
```
If I run `C:\Library> dotnet ef migrations add Initial --startup-project ../AspNetWebApp`, I need to go build "AspNetWebApp"

File: src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs (hunk 6 above)
Comment (natemcmaster) [inline src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs]: Project is not necessarily the same as startupProject. I'm going to clean this code up to make it a little more clear.

File: src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs (hunk 4 above)
Comment (natemcmaster) [inline src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs]: :cry: but I like these extension methods...

File: src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs (hunk 7 above)
Comment (NTaylorMullen) [inline src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs]: Why build them both separately though? Can you not make the assumption that the startup project references the project that will have the migrations done?

File: src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs (hunk 5 above)
Comment (natemcmaster) [inline src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs]: Build errors should be show on stderr, but I'll add the name of the project we were trying to build as well.

File: src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs (hunk 9 above)
Comment (natemcmaster) [inline src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs]: Actually, this commit didn't include a change I'm about to push. In some cases, I need AssemblyLoadContext. In others Assembly.Load is fine. AssemblyLoader is just a wrapper to make this abstraction easier.

File: src/Microsoft.EntityFrameworkCore.Tools.Cli/Program.cs (hunk 12 above)
Comment (natemcmaster) [inline src/Microsoft.EntityFrameworkCore.Tools.Cli/Program.cs]: This is by design. We only log the stack trace if we get an unexpected error type. OperationException is an exception we throw intentionally to short-circuit the command execution

File: src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs (hunk 7 above)
Comment (natemcmaster) [inline src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs]: In this scenario, AspNetWebApp references Library. The dispatcher only builds Library. We could make dispatcher smarter and have it build AspNetWebApp too, but this mixes dispatcher with inside-man too much IMO

File: src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs (hunk 10 above)
Comment (natemcmaster) [inline src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs]: Yeah, I need to cleanup the framework logic...will add in a follow-up commit.

File: src/Microsoft.EntityFrameworkCore.Tools/Program.cs (hunk 13 above)
Comment (natemcmaster) [inline src/Microsoft.EntityFrameworkCore.Tools/Program.cs]: We'd like both of these to work: `dotnet ef -v` and `dotnet -v ef`. Does dotnet utils have an API for this now?

Comment (NTaylorMullen) [inline src/Microsoft.EntityFrameworkCore.Tools/Program.cs]: Not that I know of

**Comment by natemcmaster** [comment]
I've added a new commit to this PR. It includes some response to feedback, cleanup, and fixed up some bugs.

https://github.com/aspnet/EntityFramework/pull/5142/commits/27f6fa984e2936e6b0ba896322a63cf65fe75e46

File: src/Microsoft.EntityFrameworkCore.Tools.Cli/ExecuteCommand.cs [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+                DispatchVersionOptionTemplate, GetVersion(),
+                noBuild ? NoBuildOptionTemplate : string.Empty,
+                verbose ? VerboseOptionTemplate : string.Empty
+            };
+
+        public static CommandLineApplication Create()
+        {
+            var app = new CommandLineApplication()
+            {
+                Name = "Microsoft.EntityFrameworkCore.Tools.Cli",
```
Comment (NTaylorMullen) [inline src/Microsoft.EntityFrameworkCore.Tools.Cli/ExecuteCommand.cs]: If this is tied to this assembly you should use a typeof to ensure it stays consistent without needing to manually update.

File: test/Microsoft.EntityFrameworkCore.Tools.FunctionalTests/TestUtilities/AppDomainOperationExecutor.cs [hunk 16]
```diff
@@ -118,9 +118,9 @@ private object InvokeOperationImpl(string operation, Hashtable arguments, bool i
 
             if (resultHandler.ErrorType != null)
             {
-                throw new OperationErrorException(
-                    resultHandler.ErrorType,
-                    resultHandler.ErrorStackTrace,
+                throw new OperationException(
+                    //resultHandler.ErrorType,
+                    //resultHandler.ErrorStackTrace,
```
Comment (lajones) [inline test/Microsoft.EntityFrameworkCore.Tools.FunctionalTests/TestUtilities/AppDomainOperationExecutor.cs]: TODO or remove completely?

**Comment by lajones** [comment]
:shipit:

File: src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs [hunk 17]
```diff
@@ -31,92 +33,115 @@ public class OperationExecutor
...
-                Reporter.Verbose.WriteLine("Build started...".Bold().Black());
+                Reporter.Verbose.WriteLine(ToolsCliStrings.LogBuildFailed.Bold().Black());
 
-                var buildExitCode = BuildCommandFactory.Create(startupProject,
+                var buildExitCode = BuildCommandFactory.Create(startupProjectPath,
                         startupConfiguration,
                         startupFramework,
-                        startupBuildBaseDir,
-                        startupOutputDir)
+                        buildBasePath: null,
```
Comment (NTaylorMullen) [inline src/Microsoft.EntityFrameworkCore.Tools.Cli/OperationExecutor.cs]: Might be good to allow a user to specify some of these so they don't get build output spewed in their current directory.

**Comment by NTaylorMullen** [comment]
Took a look at the CLI side of things and seemed alright. Lots here though and I'm not 100% fluent in what the EF tooling is responsible for.

:shipit: Feel free to bug me if you want anything specific looked over in detail.

**Pull request merged**